src/main/scala/com/mongodb/spark/rdd/partitioner/MongoPaginationPartitioner.scala

```diff
@@ -30,29 +31,37 @@ private[partitioner] trait MongoPaginationPartitioner {
   private implicit object BsonValueOrdering extends BsonValueOrdering
 
   protected def calculatePartitions(connector: MongoConnector, readConfig: ReadConfig, partitionKey: String, count: Long,
-                                    numDocumentsPerPartition: Int): Seq[BsonValue] = {
+                                    numDocumentsPerPartition: Int, matchQuery: Bson): Seq[BsonValue] = {
 
     @tailrec
     def accumulatePartitions(results: List[BsonValue], position: Int): List[BsonValue] = {
-      position >= count match {
-        case true => results
-        case false =>
-          val (skipValue, preBsonValue) = results.isEmpty match {
-            case true  => (0, new BsonMinKey())
-            case false => (numDocumentsPerPartition, results.head)
-          }
-          val newHead: Option[BsonValue] = connector.withCollectionDo(readConfig, { coll: MongoCollection[BsonDocument] =>
-            Option(coll.find()
-              .filter(Filters.gte(partitionKey, preBsonValue))
-              .skip(skipValue)
-              .projection(Projections.include(partitionKey))
-              .sort(Sorts.ascending(partitionKey))
-              .first()).map(doc => doc.get(partitionKey))
-          })
-          newHead.isDefined match {
-            case true  => accumulatePartitions(newHead.get :: results, position + numDocumentsPerPartition)
-            case false => results
+      if (position > count) {
```

Comment:
The trait for the Pagination based  partitioners now uses the `$match` in the filter as well as the previous maximum value or  `BsonMinKey()` if the first page.